### ISSUE TITLE
批量工作：把 Python 端的逻辑尽可能地迁移到 C++ 端

### ISSUE THREAD

**Issue by doombeaker** [Oneflow-Inc/oneflow#6156]
之前静态图接口的时代，不少算子是由基础算子在 Python 端拼接的。
这在静态图中影响不大，因为 Python 端的逻辑只会在编译图时执行一次。

但是动态图模式下，复杂的逻辑放在 Python 端会对性能造成比较显著的影响，尤其是小 batch 的情况下（计算的时间无法被其它操作覆盖）。

因此，我们需要尽可能地将逻辑，迁移到 C++ 下。目前， @BBuf  已经迁移了一部分：https://github.com/Oneflow-Inc/oneflow/pull/6114/

剩余的需要工作量依然很大，而且迁移时要注意的重点情况不一。需要大家协同参与。

这个 issue 将类似 OP 开发那样的记录方式，对算子迁移工作做计划和分工。

本主贴将记录每周的计划任务和分工。一楼跟贴是手工统计的要做的总任务。下周正式分工前会做一次讨论和基础的培训。

**Comment by doombeaker** [comment]
| 算子                                                         | 认领人 | PR                                                           | 备注                                                         |
| ------------------------------------------------------------ | ------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| PReLU                                                        |        |                                                              | 比较简单，只有一个 assert 要放到 C++ 里                      |
| ReLU                                                         |        |                                                              | 比较简单，只是inplace的check_innplace_valid应该放到C++（C++里可能已经做过了，Python层可以直接去掉，需要认领人自己确认下） |
| ReLU6                                                        |        |                                                              | 加入inplace的支持，inplace的判断放C++里                      |
| ELU                                                          |        |                                                              | 加入inplace的支持，inplace的判断放C++里                      |
| Hardsigmoid                                                  | 姚迟   | [新建标签页 (github.com)](https://github.com/Oneflow-Inc/oneflow/pull/6137) | 添加了 inplace，正在合并                                     |
| Softmax                                                      |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| softmax_op                                                   |        |                                                              | flow._C.softmax 重构完成后，这个函数就不需要了               |
| LogSoftmax                                                   |        |                                                              | 多个算子的拼接放到C++里。且确认下 __setstate__ 是否多余      |
| LogSigmoid                                                   |        |                                                              | 比较简单，搬到C++里拼接                                      |
| Softplus                                                     |        |                                                              | 比较简单，搬到C++里拼接                                      |
| Hardswish                                                    |        |                                                              | 加入inplace的支持，inplace的判断放C++里                      |
| Hardtanh                                                     |        |                                                              | 加入inplace的支持，inplace的判断放C++里                      |
| LeakyReLU                                                    |        |                                                              | 加入inplace的支持，inplace的判断放C++里                      |
| Mish                                                         |        |                                                              | 加入inplace的支持，inplace的判断放C++里。现有的inplace参数没有判断和传递，是有问题的 |
| SiLU                                                         |        |                                                              | 加入inplace的支持，inplace的判断放C++里。现有的inplace参数没有判断和传递，是有问题的 |
| SELU                                                         |        |                                                              | 加入inplace的支持，inplace的判断放C++里。现有的inplace参数没有判断和传递，是有问题的 |
| Softsign                                                     |        |                                                              | 加入inplace的支持，inplace的判断放C++里。现有的inplace参数没有判断和传递，是有问题的 |
| arange_op                                                    |        |                                                              | 把多个算子的拼接放到 C++ 里。有与 consistent tensor 有关的代码 |
| Argmax                                                       |        |                                                              | 把多个算子的拼接放到 C++ 里。                                |
| argmax_op                                                    |        |                                                              | 逻辑都迁移到C++里，这个函数就不需要了，直接导出为 oneflow.argmax 即可 |
| Argsort                                                      |        |                                                              | 把多个算子的拼接放到 C++ 里。                                |
| argsort_op                                                   |        |                                                              | 逻辑都迁移到C++里，这个函数就不需要了，直接导出为 oneflow.argmax 即可 |
| argwhere_op                                                  |        |                                                              | 把多个算子的拼接放到 C++ 里。改写过程中询问下 consistent 能否解决、lazy问题能否解决 |
| argwhere_tensor_op                                           |        |                                                              | 和 argwhere_op 一起简化                                      |
| _NormBase、_BatchNorm、BatchNorm1d、BatchNorm2d、BatchNorm3d、 |        |                                                              | 逻辑比较多，类关系也有依赖。这几个最好一起改。把逻辑放到C++里。 |
| bmm_op、bmm_op_tensor                                        |        |                                                              | 把多个算子的拼接放到 C++ 里。届时多余的函数去掉              |
| BroadCastLike、broadcast_like_op                             |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。使用的 _calc_broadcast_axes 的逻辑也要放到 C++ 里 |
| chunk_op                                                     |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| concat_op                                                    |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| ToConsistent、to_consistent_op                               |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。涉及到 consistent |
| ToLocal                                                      |        |                                                              | 这个类有点奇怪，应该是不需要的，可以直接调用 `flow._C.to_local` |
| to_local_op                                                  |        |                                                              | assert 搬到C++里                                             |
| _ConstantBase、Ones、ones_op、Zeros、zeros_op、Full、full_op、NewOnes、new_ones_op |        |                                                              | 均在 constant.py 中，类有继承关系，最好由一个人一起完成。有些类其实是没必要存在的了，届时可以直接调用 `flow._C.xxx` |
| container.py                                                 |        |                                                              | 貌似不用搬运，暂略过                                         |
| ConvUtil、Conv1d、Conv2d、Conv3d、ConvTranspose2d            |        |                                                              | 均在 conv.py 文件中，类之间有依赖关系。逻辑比较复杂。应该是1~3个人，比较紧密合作的一起把这个文件给搬运了 |
| dataset.py                                                   |        |                                                              | 这个文件里的多个 OFrecord 有关的Reader，量多，但是逻辑比较简单。可以由一个人搬运 |
| empty_op                                                     |        |                                                              | 涉及到 sbp、placement 的判断，python代码多但是逻辑不复杂，不确定是否要（能）搬到C++ |
| expand_op                                                    |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| eye_op                                                       |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。与 ones、zeros、diag、cat 有关系 |
| _flow_flatten                                                |        |                                                              | 可以直接去掉，Tensor 方法由 `_C.flatten` 导入，举手之劳      |
| flip_op                                                      |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| floor_op、floor_op_tensor                                    |        |                                                              | 应该可以删除，改为直接导出了                                 |
| gather_op                                                    |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| gather_nd_op                                                 |        |                                                              | 应该可以删除，改为直接导出了                                 |
| InTopk、in_top_k_op、in_top_k_op_tensor                      |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| index_select_op、index_select_op_tensor                      |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| _InstanceNorm、InstanceNorm1d、InstanceNorm2d、InstanceNorm3d |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。注意那些用于小 utils 函数，也要搬到C++里做 |
| Interpolate、interpolate                                     |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| Linear                                                       |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| logical_ops.py                                               |        |                                                              | 里面有 and、xor 等逻辑运算，都比较简单，可以一起搬运。       |
| L1Loss                                                       |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| CrossEntropyLoss                                             |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| BCELoss                                                      |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| NLLLoss                                                      |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| KLDivLoss                                                    |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| MSELoss                                                      |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| MarginRankingLoss                                            |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| CTCLoss                                                      |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| BCEWithLogitsLoss                                            |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| SmoothL1Loss                                                 |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| CombinedMarginLoss                                           |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| masked_fill_op                                               |        |                                                              | 貌似不用（好）搬，完美解决方案可能是直接改 kernel，增加支持的index类型 |
| masked_select_op、tensor_masked_select_op                    |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| variance_op                                                  |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| sin_op                                                       |        |                                                              | 应该可以删除，改为直接导出了                                 |
| sin_op_tensor                                                |        |                                                              | 应该可以删除，改为直接导出了                                 |
| std_op                                                       |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| addmm、addmm_op                                              |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。不过即使不搬，目前这两个，好像只需要留一个就可以了。 |
| Clamp                                                        |        |                                                              | 可能不用（能）搬，待定                                       |
| Topk、topk_op                                                |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| matmul_op                                                    |        |                                                              | 应该可以删除，改为直接导出了                                 |
| MeshGrid、meshgrid_op                                        |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| Narrow、narrow_op                                            |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| nonzero_op、nonzero_tensor_op                                |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| Vector_Norm、vector_norm_tensor_op                           |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| Matrix_Norm、matrix_norm_tensor_op                           |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| Norm、norm_op、norm_tensor_op                                |        |                                                              | 把多个算子的拼接放到 C++ 里，届时多余的函数去掉。            |
| GroupNorm                                                    |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| LayerNorm                                                    |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| one_hot                                                      |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| ReplicationPad2d                                             |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| ReflectionPad2d                                              |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| ConstantPad1d                                                |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| ConstantPad2d                                                |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| ConstantPad3d                                                |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| ZeroPad2d                                                    |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
|                                                              |        |                                                              | 以上pad 类型的module，保留了 builtin_op，不确认可以搬运成functor，需要再评估 |
| permute_op                                                   |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| PixelShufflev2                                               |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| MaxPool1d、MaxPool2d、MaxPool3d                              |        |                                                              | 把多个算子的拼接放到 C++ 里                                  |
| AdaptiveAvgPool1d、adaptive_avg_pool1d、AdaptiveAvgPool2d、adaptive_avg_pool2d、AdaptiveAvgPool3d、adaptive_avg_pool3d |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| Rand、rand_op                                                |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| RandN、randn_op                                              |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| RandInt、randint_op                                          |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| RandPerm、randperm_op                                        |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| reduce_ops.py 里的max、min、sum等操作                        |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数。比较简单，可以一起做。 |
| relu.py                                                      |        |                                                              | 应该可以删除，改为直接导出了                                 |
| repeat_op                                                    |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| reshape_op                                                   |        |                                                              | 应该可以删除，改为直接导出了                                 |
| reshape_tensor_op                                            |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| view_op                                                      |        |                                                              | 涉及底层的改动，可以暂时跳过。等待 stride 机制等             |
| scatter、scatter_add、_scatter_nd_op                         |        |                                                              | 应该可以删除，改为直接导出了                                 |
| slice_op                                                     |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| slice_update_op                                              |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| logical_slice_assign_op                                      |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| Sort、sort_op                                                |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| Embedding                                                    |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| Split、split_op                                              |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| squeeze_op                                                   |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| stack                                                        |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| tensor_buffer.py                                             |        |                                                              | 不确定是否能搬，待确认                                       |
| negative_op                                                  |        |                                                              | 应该可以删除，改为直接导出了                                 |
| tensor_ops.py                                                |        |                                                              | 里面的方法都比较简单和特殊，需要进一步确认是否需要搬到 C++   |
| tile_op                                                      |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| to.py                                                        |        |                                                              | 文件中的方法比较特殊，需要进一步确认                         |
| transpose_op                                                 |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| Atan2                                                        |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| unsqueeze_op                                                 |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
| where_op                                                     |        |                                                              | 把多个算子的拼接放到 C++ 里。届时去掉多余的函数              |
|                                                              |        |                                                              |                                                              |
|                                                              |        |                                                              |                                                              |
|                                                              |        |                                                              |                                                              |
|                                                              |        |                                                              |                                                              |
|                                                              |        |                                                              |                                                              |
|                                                              |        |                                                              |                                                              |
|                                                              |        |                                                              |                                                              |

**Comment by yuanms2** [comment]
有证据表明这个问题有明星影响吗

**Comment by doombeaker** [comment]
> 有证据表明这个问题有明显影响吗

这个我听新奇和我提过。可能新奇和后江最清楚细节和证据。 @lixinqi @hjchen2

**Comment by hjchen2** [comment]
可以找一两个模型，把那些python逻辑比较重的接口直接替换成flow._C的接口先试试。indexing操作目前是一个瓶颈，下周出一个view的版本看看效果。

**Comment by wyg1997** [comment]
补充一点，有一部分 op 有 in-place 版本和 out-place 两种版本，在 torch 中的表现是有 xxx_op() 和 xxx_op_() 两个接口。这些 op 要在 functional 中导出 `oneflow._C.xxx_op()` 和 `oneflow._C.xxx_op_()` 两个接口（**而不是**导出 oneflow._C.xxx_op(x, inplace=False)这样的接口），然后用 add_docs（可以参考 oneflow/framework/docstr/math_ops.py 中添加文档的方式）来添加对应的文档。

总体的原则就是：**尽量少写 python wrapper，使 oneflow._C 导出的接口尽可能的对齐 torch 的接口**
